nucliadb/nucliadb/search/search/chat/prompt.py

```diff
@@ -154,3 +162,207 @@ async def get_chat_prompt_context(
+    """
+    # Collect the list of resources in the results (in order of relevance).
+    ordered_paras = get_ordered_paragraphs(results)
+    ordered_resources = []
+    for paragraph in ordered_paras:
+        resource_uuid = paragraph.id.split("/")[0]
+        if resource_uuid not in ordered_resources:
+            ordered_resources.append(resource_uuid)
+
+    # Fetch the extracted texts of the specified fields for each resource
+    txn = await get_read_only_transaction()
+    kb_obj = KnowledgeBoxORM(txn, await get_storage(), kbid)
+
+    tasks = [
+        get_resource_field_extracted_text(kb_obj, resource_uuid, field_id)
+        for resource_uuid in ordered_resources
+        for field_id in extend_with_fields
+    ]
+    field_extracted_texts = await run_concurrently(tasks)
+
+    output: Context = {}
+    for result in field_extracted_texts:
+        if result is None:
+            continue
+        # Add the extracted text of each field to the beginning of the context.
+        field, extracted_text = result
+        output[field.resource_unique_id] = extracted_text
+
+    # Add the extracted text of each paragraph to the end of the context.
+    for paragraph in ordered_paras:
+        output[paragraph.id] = _clean_paragraph_text(paragraph)
+    return output
+
+
+class PromptContextBuilder:
+    """
+    Builds the context for the LLM prompt.
+    """
+
+    def __init__(
+        self,
+        kbid: str,
+        find_results: KnowledgeboxFindResults,
+        user_context: Optional[list[str]] = None,
+        options: Optional[RAGOptions] = None,
+    ):
+        self.kbid = kbid
+        self.find_results = find_results
+        self.user_context = user_context
+        self.options = options
+
+    def prepend_user_context(self, context: Context) -> Context:
+        # Chat extra context passed by the user is the most important, therefore
+        # it is added first, followed by the found text blocks in order of relevance
+        extended = {
+            f"USER_CONTEXT_{i}": text_block
+            for i, text_block in enumerate(self.user_context or [])
+        }
+        extended.update(context)
+        return extended
+
+    async def build(self) -> tuple[Context, ContextOrder]:
+        context = await self._build_context()
+        context = self.prepend_user_context(context)
+        context_order = {
+            text_block_id: order for order, text_block_id in enumerate(context.keys())
+        }
+        return context, context_order
+
+    async def _build_context(self):
+        if self.options is None or self.options.context_strategies is None:
+            return await default_prompt_context(self.kbid, self.find_results)
+
+        chosen_strategies = self.options.context_strategies
+        if ContextStrategy.FULL_RESOURCE in chosen_strategies:
+            return await full_resource_prompt_context(self.kbid, self.find_results)
+
+        return await composed_prompt_context(
```

Comment:
Feels kind of weird how `extend_with_fields` implicitly activates the strategy even though could not be in your chosen strategies. Maybe a different structure in `RagOption` or to explicitly check for the strategy will solve it